documentation/correctness.adoc

```diff
@@ -0,0 +1,97 @@
+
+LionWeb provides a precise notion for model correctness in the form of a multi-leveled specification.
+For each level, it specifies what kinds of incorrectness can arise, how these should be reported as issues, and whether and how these can be meaningfully _recovered_ from.
+This approach provides a degree of resilience: issues arising in one level don't necessarily prohibit determining whether a model exhibits incorrectness on another level, provided an issue can be meaningfully _recovered_ from.
+
+LionWeb provides a JSON serialization format for models: the serialization chunk.
+LionWeb intentionally doesn't prescribe anything about the _programmatic representation_ of models -- i.e.: the runtime or in-memory representation that's the result of deserializing a serialization chunk of a model -- leaving implementors of language-oriented tooling free to choose the representation that's right for them.
+Because of that, this specification is phrased _completely_ in terms of the JSON serialization.
+The first three levels of this specification pertain purely to the JSON that's supposed to be the serialization chunk of a model, and are language(s)-aspecific.
+The next levels _could_ be phrased in terms of a programmatic representation, but the equivalent formulation purely in terms of the serialization format works just as well.
+
+This specification admits some flexibility on the first three, JSON-centric levels.
+That is because the behavior and adaptability of JSON parsers differ from implementation to implementation.
+We simply can't (and don't want to demand) that a LionWeb implementation always produces issues, even if they're recoverable or ignorable anyway.
+
+
+== Overview of the levels
+
+We first give an overview of the various levels, and what each pertains to, at first without specifying the corresponding kinds of issues that may arise in that level.
+The list is roughly in order from low-level to language(s)-specific.
+
+[horizontal]
+
+Low-level syntactic::
+Pertains to the character stream that's supposed to be JSON text that's (supposed to be) the serialization chunk of a model.
+
+Structural syntactic::
+Pertains to the syntactic structure of the parsed JSON text.
+It's essentially equivalent to applying a generic JSON Schema.
+
+Hierarchical::
+Pertains to non-local, inter-node constraints on the parsed JSON text.
+An example is that a node that declared a parent node must be contained as a child of that parent node.
+
+Meta-structural::
+This means that the serialization chunk conforms to the language(s) it declares to be an instance of.
+An example is that a node must declare values for exactly the features of the classifier it declares.
+
+Referential::
+This pertains to whether references resolve.
+An unresolvable reference doesn't prevent being able to work with the model, but it often prohibits the next level –– Constraints -- from producing meaningful results.
+
+Constraints::
```

Comment:
Does this include or exclude scoping and typing?

Reply:
Scoping doesn't play so much of a role within a chunk because of explicit (ID-based) references. Of course, the existence of a scoping rule implies the existence of a constraint – e.g., “the field referenced should be a field of the entity referenced by the parent”, so in that sense scoping is included.

Typing is very much included. I think very roughly ~50% of all constraints would be directly “type-informed” anyway.